<launch>
  <arg name="ODOM_FRAME" default="odom" />
  <group ns="laser_preprocess">
    <node pkg="nodelet" type="nodelet" name="laser_preprocess_manager"
          args="manager" output="screen"/>
    <node pkg="nodelet" type="nodelet"
          name="tilt_laser_listener"
          args="load jsk_pcl/TiltLaserListener laser_preprocess_manager"
          machine="vmachine"
          output="screen">
      <remap from="~input" to="/joint_states" />
      <rosparam>
        laser_type: infinite_spindle_half
        joint_name: motor_joint
        use_laser_assembler: true
        overwrap_angle: 0.1
      </rosparam>
      <remap from="~assemble_scans2" to="/laser_preprocess/assemble_scans2" />
    </node>

    <node pkg="laser_assembler" type="laser_scan_assembler"
          name="laser_scan_assembler_node"
          machine="vmachine"
          output="screen" >
      <remap from="scan" to="/multisense/lidar_scan_filtered"/>
      <param name="tf_cache_time_secs" value="20" />
      <param name="max_scans" type="int" value="1000" />
      <param name="fixed_frame" value="$(arg ODOM_FRAME)" />
    </node>

    <node pkg="nodelet" type="nodelet"
          name="downsampler"
          machine="vmachine"
          args="load pcl/VoxelGrid laser_preprocess_manager">
      <remap from="~input" to="tilt_laser_listener/output_cloud" />
      <rosparam>
        leaf_size: 0.02
        filter_limit_min: -100
        filter_limit_max: 100
      </rosparam>
    </node>
    <node pkg="nodelet" type="nodelet"
          name="gsensor_cloud"
          machine="vmachine"
          args="load jsk_pcl/TfTransformCloud laser_preprocess_manager">
      <remap from="~input" to="downsampler/output" />
      <rosparam>
        target_frame_id: ground
      </rosparam>
    </node>

    <node pkg="nodelet" type="nodelet"
          name="x_filter"
          machine="vmachine"
          args="load pcl/PassThrough laser_preprocess_manager">
      <remap from="~input" to="gsensor_cloud/output" />
      <rosparam>
        filter_field_name: x
        filter_limit_min: -10
        filter_limit_max: 10
        filter_limit_negative: false
      </rosparam>
    </node>
    <node pkg="nodelet" type="nodelet"
          name="y_filter"
          machine="vmachine"
          args="load pcl/PassThrough laser_preprocess_manager">
      <remap from="~input" to="x_filter/output" />
      <rosparam>
        filter_field_name: y
        filter_limit_min: -10
        filter_limit_max: 10
        filter_limit_negative: false
      </rosparam>
    </node>

    <!-- first, remove upper pointcloud -->
    <node pkg="nodelet" type="nodelet"
          name="z_filter"
          machine="vmachine"
          args="load pcl/PassThrough laser_preprocess_manager">
      <remap from="~input" to="y_filter/output" />
      <rosparam>
        filter_field_name: z
        filter_limit_negative: true
        filter_limit_min: 3.0
        filter_limit_max: 100.0
      </rosparam>
    </node>

    <node pkg="nodelet" type="nodelet"
          name="odom_cloud"
          machine="vmachine"
          args="load jsk_pcl/TfTransformCloud laser_preprocess_manager">
      <remap from="~input" to="z_filter/output" />
      <rosparam>
        target_frame_id: odom
      </rosparam>
    </node>
  </group>
</launch>
